### PR TITLE
fix: handle errors appropriately as they are wrapped

### PR DESCRIPTION
**Pull request by harshavardhana**
## Description
fix: handle errors appropriately as they are wrapped

## Motivation and Context
use error checks which unwraps errors properly

## How to test this PR?
nothing special

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Documentation needed
- [ ] Unit tests needed

### CODE REVIEW

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-zoned.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
Deleting image on docker hub
Deleting image locally

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-zoned.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :x: [more...](#fa7ac6b6-da6a-4274-9239-3de5aee6e606) |
<a name="fa7ac6b6-da6a-4274-9239-3de5aee6e606"></a>
### 10917-c47d9a2/mint-gateway-azure.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev8.minio.io:30727
ACCESS_KEY:           minioazure
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 43667c45d4c3:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 9 seconds
(2/15) Running aws-sdk-java tests ... done in 1 seconds
(3/15) Running aws-sdk-php tests ... done in 2 minutes and 15 seconds
(4/15) Running aws-sdk-ruby tests ... done in 20 seconds
(5/15) Running awscli tests ... done in 3 minutes and 19 seconds
(6/15) Running healthcheck tests ... done in 0 seconds
(7/15) Running mc tests ... done in 3 minutes and 42 seconds
(8/15) Running minio-dotnet tests ... done in 1 minutes and 42 seconds
(9/15) Running minio-go tests ... done in 6 minutes and 23 seconds
(10/15) Running minio-java tests ... done in 11 minutes and 4 seconds
(11/15) Running minio-js tests ... done in 2 minutes and 39 seconds
(12/15) Running minio-py tests ... done in 19 minutes and 54 seconds
(13/15) Running s3cmd tests ... FAILED in 1 minutes and 20 seconds
{
  "name": "s3cmd",
  "duration": "18340",
  "function": "test_get_object_multipart",
  "status": "FAIL",
  "error": "\n!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!\n    An unexpected error has occurred.\n  Please try reproducing the error using\n  the latest s3cmd code from the git master\n  branch found at:\n    https://github.com/s3tools/s3cmd\n  and have a look at the known issues list:\n    https://github.com/s3tools/s3cmd/wiki/Common-known-issues-and-their-solutions\n  If the error persists, please report the\n  following lines (removing any private\n  info as necessary) to:\n   s3tools-bugs@lists.sourceforge.net\n\n\n!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!\n\nInvoked as: /usr/local/bin/s3cmd --config /tmp/.s3cmd-20457/s3cfg get s3://s3cmd-test-bucket-25875/s3cmd-test-object-29955 s3cmd-test-object-29955.downloaded\nProblem: <class 'ConnectionResetError: [Errno 104] Connection reset by peer\nS3cmd:   2.1.0\npython:   3.6.9 (default, Jul 17 2020, 12:50:27) \n[GCC 8.4.0]\nenvironment LANG=C.UTF-8\n\nTraceback (most recent call last):\n  File \"/usr/local/bin/s3cmd\", line 3121, in <module>\n    rc = main()\n  File \"/usr/local/bin/s3cmd\", line 3030, in main\n    rc = cmd_func(args)\n  File \"/usr/local/bin/s3cmd\", line 604, in cmd_object_get\n    response = s3.object_get(uri, dst_stream, destination, start_position = start_position, extra_label = seq_label)\n  File \"/usr/local/lib/python3.6/dist-packages/S3/S3.py\", line 726, in object_get\n    response = self.recv_file(request, stream, labels, start_position)\n  File \"/usr/local/lib/python3.6/dist-packages/S3/S3.py\", line 1733, in recv_file\n    data = http_response.read(this_chunk)\n  File \"/usr/lib/python3.6/http/client.py\", line 459, in read\n    n = self.readinto(b)\n  File \"/usr/lib/python3.6/http/client.py\", line 503, in readinto\n    n = self.fp.readinto(b)\n  File \"/usr/lib/python3.6/socket.py\", line 586, in readinto\n    return self._sock.recv_into(b)\nConnectionResetError: [Errno 104] Connection reset by peer\n\n!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!\n    An unexpected error has occurred.\n  Please try reproducing the error using\n  the latest s3cmd code from the git master\n  branch found at:\n    https://github.com/s3tools/s3cmd\n  and have a look at the known issues list:\n    https://github.com/s3tools/s3cmd/wiki/Common-known-issues-and-their-solutions\n  If the error persists, please report the\n  above lines (removing any private\n  info as necessary) to:\n   s3tools-bugs@lists.sourceforge.net\n!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!"
}
(13/15) Running s3select tests ... done in 1 minutes and 11 seconds
(14/15) Running security tests ... done in 0 seconds

Executed 14 out of 15 tests successfully.
```
Deleting image on docker hub
Deleting image locally

**Review by klauspost — CHANGES_REQUESTED** [review]
You cannot safely do this. The first block is not written until it has actual data (5k entries) this can take an arbitrary amount of time depending on system load.

That is why we check if there hasn't been updates for a while, only then will the cache become invalid.

I considered rebuilding instead of failing, but rejected it since it leaves the cache inconsistent. The master does not know that the cache is being rebuilt and you will suddenly have races trying to rebuild the same cache with blocks potentially being written by different list operations.

In this for instance you have no control of whether 5 list operations will start rebuilding the same cache.

**Comment by harshavardhana** [comment]
> You cannot safely do this. The first block is not written until it has actual data (5k entries) this can take an arbitrary amount of time depending on system load.
> 
> That is why we check if there hasn't been updates for a while, only then will the cache become invalid.
> 
> I considered rebuilding instead of failing, but rejected it since it leaves the cache inconsistent. The master does not know that the cache is being rebuilt and you will suddenly have races trying to rebuild the same cache with blocks potentially being written by different list operations.
> 
> In this for instance you have no control of whether 5 list operations will start rebuilding the same cache.

Can we not hold a lock and do it @klauspost ? because this looks like a potential problem right now - is this also the reason why listing blocks for concurrent calls? as per the slow listing issue?

**Comment by klauspost** [comment]
> Now lets say if .metacache is purged or missing somehow

This should be a relatively uncommon scenario. It would "require" a hard system crash combined with the save operation corrupting the existing data.

Either way, if the list is unknown to the cache manager it will be re-created by the first requester.

**Comment by klauspost** [comment]
> Can we not hold a lock

It will make no practical difference if we do it lockless or with locks. The time needed to list will be the same and in case if a dead server the lock will need to time out anyway, so you are in the same situation.

>  is this also the reason why listing blocks for concurrent calls

It blocks, because it is designed to not list the same paths concurrently. The second call will need the first block to have been written to be able to return data.

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-zoned.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :x: [more...](#e2797b30-8f14-4b17-90de-9ff97041f46f) |
<a name="e2797b30-8f14-4b17-90de-9ff97041f46f"></a>
### 10917-f869b0f/mint-gateway-azure.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev6.minio.io:31828
ACCESS_KEY:           minioazure
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp d4202c3018df:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 8 seconds
(2/15) Running aws-sdk-java tests ... done in 2 seconds
(3/15) Running aws-sdk-php tests ... done in 2 minutes and 18 seconds
(4/15) Running aws-sdk-ruby tests ... done in 19 seconds
(5/15) Running awscli tests ... done in 2 minutes and 52 seconds
(6/15) Running healthcheck tests ... done in 0 seconds
(7/15) Running mc tests ... done in 3 minutes and 40 seconds
(8/15) Running minio-dotnet tests ... done in 1 minutes and 41 seconds
(9/15) Running minio-go tests ... done in 8 minutes and 23 seconds
(10/15) Running minio-java tests ... done in 11 minutes and 4 seconds
(11/15) Running minio-js tests ... done in 2 minutes and 40 seconds
(12/15) Running minio-py tests ... done in 18 minutes and 46 seconds
(13/15) Running s3cmd tests ... done in 2 minutes and 20 seconds
(14/15) Running s3select tests ... FAILED in 6 seconds
{
  "name": "s3select:test_csv_input_quote_char",
  "function": "select_object_content(bucket_name, object_name, opts)",
  "args": {
    "bucket_name": "s3select-test-0d543393-82b3-496e-bff0-d89080401bac"
  },
  "duration": 5812,
  "message": "Test test_6 unexpectedly failed with: 'event-type'",
  "error": "Traceback (most recent call last):\n  File \"/mint/run/core/s3select/csv.py\", line 39, in test_sql_api\n    for d in data.stream(10*1024):\n  File \"/usr/local/lib/python3.6/dist-packages/minio/select/reader.py\", line 211, in stream\n    message = self.__extract_message()\n  File \"/usr/local/lib/python3.6/dist-packages/minio/select/reader.py\", line 161, in __extract_message\n    event_type = header_map[\"event-type\"]\nKeyError: 'event-type'\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"./tests.py\", line 52, in main\n    test_csv_input_custom_quote_char(client, log_output)\n  File \"/mint/run/core/s3select/csv.py\", line 119, in test_csv_input_custom_quote_char\n    data, sql_opts, expected_output)\n  File \"/mint/run/core/s3select/csv.py\", line 45, in test_sql_api\n    'Test {} unexpectedly failed with: {}'.format(test_name, select_err))\nValueError: Test test_6 unexpectedly failed with: 'event-type'\n",
  "status": "FAIL"
}
(14/15) Running security tests ... done in 0 seconds

Executed 14 out of 15 tests successfully.
```
Deleting image on docker hub
Deleting image locally

**Comment by harshavardhana** [comment]
> It blocks, because it is designed to not list the same paths concurrently. The second call will need the first block to have been written to be able to return data.

But the potential problem here is that it invariably affects most common scenarios - concurrent listing is actually quite common @klauspost

**Comment by klauspost** [comment]
>  concurrent listing is actually quite common

@harshavardhana That is why they share the cache instead of adding more load to the disks by requesting data that is already being generated or available. That is the whole idea, to share the listings across calls instead of cold generating them for every request.

**Comment by harshavardhana** [comment]
> @harshavardhana That is why they share the cache instead of adding more load to the disks by requesting data that is already being generated or available. That is the whole idea, to share the listings across calls instead of cold generating them for every request.

@klauspost correct I am just worried that it affects some common scenarios and it seems to affect on a regular basis.

**Comment by harshavardhana** [comment]
Going to change this PR into more of a layered error handling change. instead of trying to recreate the cache @klauspost

**Comment by klauspost** [comment]
@harshavardhana Yes, in some cases it will slow down, main problem is there is no way to tell.

Consider this scenario:

Request A: list `prefix/`, no separator, so listing recursively .
Request B: list `prefix/folder`, any way.

In this case Request B will have to wait for the result of Request A to reach `prefix/folder` since it will be pulling results from that. This can be a longer wait than it would otherwise see, since it would be able to pull results right away. 

However, since we don't know the size of `prefix/folder` it may be a significant amount of duplicate operations if it was to disregard the cache.

Previously Request B) alone would be able to disregard a lot of results and needs only to list objects in the `prefix` folder with `folder` prefix. To have a usable cache we need to list all elements in the folder and resolve if they are objects, otherwise the cache will be useless to anyone not listing with that prefix.

So `prefix/a` and `prefix/b` will be able to share the cache, but it will require that we resolve the entire `prefix` folder. The first request will of course be able to return results as soon as the `prefix/a` has been started, since we can return results while we are listing.

We *can* put in a prefix filter into the cache itself, but it would disallow a lot of cache sharing.

**Comment by harshavardhana** [comment]
> @harshavardhana Yes, in some cases it will slow down, main problem is there is no way to tell.
> 
> Consider this scenario:
> 
> Request A: list `prefix/`, no separator, so listing recursively .
> Request B: list `prefix/folder`, any way.
> 
> In this case Request B will have to wait for the result of Request A to reach `prefix/folder` since it will be pulling results from that. This can be a longer wait than it would otherwise see, since it would be able to pull results right away.
> 
> However, since we don't know the size of `prefix/folder` it may be a significant amount of duplicate operations if it was to disregard the cache.
> 
> Previously Request B) alone would be able to disregard a lot of results and needs only to list objects in the `prefix` folder with `folder` prefix. To have a usable cache we need to list all elements in the folder and resolve if they are objects, otherwise the cache will be useless to anyone not listing with that prefix.
> 
> So `prefix/a` and `prefix/b` will be able to share the cache, but it will require that we resolve the entire `prefix` folder. The first request will of course be able to return results as soon as the `prefix/a` has been started, since we can return results while we are listing.
> 
> We _can_ put in a prefix filter into the cache itself, but it would disallow a lot of cache sharing.

I understood the problem but I am curious to see how many more reports of list slowing down during concurrent listing comes up, because it is also indicated that the issue repeats itself after a cache has sort of expired (reached end of its life)

Like in a mutating prefix and concurrency listing it can become sort of a major problem.

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-zoned.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :x: [more...](#239a078c-fe5f-4198-9487-536491c1ecd2) |
<a name="239a078c-fe5f-4198-9487-536491c1ecd2"></a>
### 10917-cb27ef9/mint-gateway-azure.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev7.minio.io:31291
ACCESS_KEY:           minioazure
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 38a9393ce011:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 9 seconds
(2/15) Running aws-sdk-java tests ... done in 1 seconds
(3/15) Running aws-sdk-php tests ... done in 2 minutes and 19 seconds
(4/15) Running aws-sdk-ruby tests ... done in 20 seconds
(5/15) Running awscli tests ... done in 2 minutes and 54 seconds
(6/15) Running healthcheck tests ... done in 0 seconds
(7/15) Running mc tests ... done in 3 minutes and 41 seconds
(8/15) Running minio-dotnet tests ... done in 1 minutes and 38 seconds
(9/15) Running minio-go tests ... done in 6 minutes and 45 seconds
(10/15) Running minio-java tests ... FAILED in 9 minutes and 57 seconds
{
  "name": "minio-java",
  "function": "composeObject(ComposeObjectArgs args)",
  "args": "size: 6 MB & 6 MB ",
  "duration": 706,
  "status": "FAIL",
  "error": "java.net.SocketException: Connection reset >>> [java.base/java.net.SocketInputStream.read(SocketInputStream.java:186), java.base/java.net.SocketInputStream.read(SocketInputStream.java:140), okio.Okio$2.read(Okio.java:140), okio.AsyncTimeout$2.read(AsyncTimeout.java:237), okio.RealBufferedSource.indexOf(RealBufferedSource.java:358), okio.RealBufferedSource.readUtf8LineStrict(RealBufferedSource.java:230), okhttp3.internal.http1.Http1ExchangeCodec.readHeaderLine(Http1ExchangeCodec.java:242), okhttp3.internal.http1.Http1ExchangeCodec.readResponseHeaders(Http1ExchangeCodec.java:213), okhttp3.internal.connection.Exchange.readResponseHeaders(Exchange.java:115), okhttp3.internal.http.CallServerInterceptor.intercept(CallServerInterceptor.java:94), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:142), okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.java:43), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:142), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:117), okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.java:94), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:142), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:117), okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.java:93), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:142), okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.java:88), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:142), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:117), okhttp3.RealCall.getResponseWithInterceptorChain(RealCall.java:229), okhttp3.RealCall.execute(RealCall.java:81), io.minio.MinioClient.execute(MinioClient.java:1014), io.minio.MinioClient.uploadPart(MinioClient.java:7781), io.minio.MinioClient.putObject(MinioClient.java:4706), io.minio.MinioClient.putObject(MinioClient.java:4881), FunctionalTest.composeObject_test1(FunctionalTest.java:2346), FunctionalTest.runTests(FunctionalTest.java:3825), FunctionalTest.main(FunctionalTest.java:4058)]"
}
(10/15) Running minio-js tests ... done in 2 minutes and 39 seconds
(11/15) Running minio-py tests ... done in 18 minutes and 48 seconds
(12/15) Running s3cmd tests ... done in 2 minutes and 29 seconds
(13/15) Running s3select tests ... done in 59 seconds
(14/15) Running security tests ... done in 0 seconds

Executed 14 out of 15 tests successfully.
```
Deleting image on docker hub
Deleting image locally

**Comment by klauspost** [comment]
> because it is also indicated that the issue repeats itself after a cache has sort of expired (reached end of its life)

Correct, it will as long as the bloom filter of the path is dirty.

> Like in a mutating prefix and concurrency listing it can become sort of a major problem.

I am not quite following. Could you give an example?

File: cmd/metacache-set.go [hunk 1]
```diff
@@ -531,14 +534,23 @@ func (er *erasureObjects) listPath(ctx context.Context, o listPathOptions) (entr
...
-		case nil, io.EOF, context.Canceled, context.DeadlineExceeded:
-			return entries, err
+		if IsErr(err, []error{
+			nil,
+			io.EOF,
+			context.Canceled,
+			context.DeadlineExceeded,
+		}...) {
+			// Expected good errors we don't need to return error.
+			return entries, nil
```
Comment (klauspost) [inline cmd/metacache-set.go:547]: `io.EOF` bears meaning and must be returned. The rest are harmless to ignore.

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-zoned.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
Deleting image on docker hub
Deleting image locally

**Comment by harshavardhana** [comment]
PTAL @vadmeste @klauspost

**Review by klauspost — APPROVED** [review]

**Pull request merged**